### ISSUE TITLE
Bug. TreeGrid collapsing/expanding functionallity freezes

### ISSUE THREAD

**Issue by drewoko** [vaadin/framework#10104]
Good day,

Met one issue with Treegrid component on Vaadin 8+ versions.

We leave in very dinamic world where data is changing very fast and we can't be sure that one query that returned one result will respond with same result after one second. 

Problem is simple. If we impementing AbstractHierarchicalDataProvider and query inside hasChildren metrod returned positive number of elements and then when user clicked on arrow to expand this element and fetchChildren returned empty stream, then tree table collapsing/expanding functionallity will be freezed.

Solution to cache data on hasChildren call is inappropriate solution, because we should cache a lot of data in memory, and we are not sure, that even it will be used by customer. And data will not be up-to-date.

[Possibly related ticket from different angle](https://github.com/vaadin/framework/issues/8492)

I'm attached zip with demonstration of that issue.
[TreeGridDemo.zip](https://github.com/vaadin/framework/files/1339807/TreeGridDemo.zip)

**Comment by joker1** [comment]
I am also affected by this bug. Collapse/expand functionality freezes for every item in TreeGrid.

It happens when hasChildrent(T item) method returns true but subsequent getChildCount() returns 0 and fetchChildren() returns empty stream.

**Comment by joker1** [comment]
I was able to workaround this bug by hard-coding "return 1;" in dataProvider for parent entries that do not actually have children. Also fetchChildren will return stream of 1 dummy item for these parent entries. Additionally, i added ExpandListener that refreshes these dataProvider items so that those dummy children will not be actually shown.

`
@Override
public int getChildCount(HierarchicalQuery<LdapTreeItem, LdapTreeFilter> query) {
	int childCount = countChildren(query.getParent());
	Boolean hasSubordinates = query.getParent().getHasSubordinates();
	if(hasSubordinates == null && childCount == 0) {
		/* We have previously returned true from hasChildren(...) so we need to return non-zero child count. */
		return 1;
	} else {
		return childCount;
	}
}

@Override
public int getChildCount(HierarchicalQuery<LdapTreeItem, LdapTreeFilter> query) {
	List<?> children = getChildren(query.getParent());
	Boolean hasSubordinates = query.getParent().getHasSubordinates();
	if(hasSubordinates == null && children.isEmpty()) {
		/* We have previously returned true from hasChildren(...) so we need to return non-empty children stream. */
		return Arrays.asList(new DummyItem()).stream();
	} else {
		return children.stream();
	}
}
`

Add listener to Tree/TreeGrid
`addExpandListener(e-> {
	T item = e.getExpandedItem();
	if (item.getHasSubordinates() == null && dataProvider.countChildren(item) == 0) {
		item.setHasSubordinates(false);
		dataProvider.refreshItem(item);
	}
});`

**Comment by TatuLund** [comment]
Is this related https://github.com/vaadin/framework/issues/10317 ?

**Comment by TatuLund** [comment]
If you are doing expandAll / collapseAll, there is new methods for this coming https://github.com/vaadin/framework/pull/10283  the patch is already in 8.4-SNAPSHOT.

**Comment by drewoko** [comment]
@TatuLund no, it's not. It's related to data consistency. If we are telling to node in treegrid, that it's have children, then user expands it and there is 0 element's given by data provider, then UI freezes.

**Comment by drewoko** [comment]
just for note. Issue prioritized using vaadin subscription.

**Comment by macjuan** [comment]
Please consider this as a BFP, just chatted with Deniss and he is not happy with workarounds.

**Comment by pleku** [comment]
>@macjuan BFP was removed?

Sorry, but his is not considered as a bug as it has been stated in the corresponding support portal issue too.

This by design how the hierarchical data provider & `TreeGrid` works in FW 8, it does not support dynamic sizes for the tree nodes. There are some ways to workaround this and `TreeGrid` could be made into working with this case, but that would require quite much changes that we have not planned at the moment.

Contributions are welcome though, if someone wants to take a look at this, the team can give some pointers and will review the changes.
